library/src/scala/IArray.scala

```diff
@@ -521,4 +601,575 @@ object IArray:
   def unapplySeq[T](x: IArray[T]): Array.UnapplySeqWrapper[_ <: T] =
     Array.unapplySeq(x)
 
-end IArray
\ No newline at end of file
+  /** A lazy filtered array. No filtering is applied until one of `foreach`, `map` or `flatMap` is called. */
+  class WithFilter[T](p: T => Boolean, xs: IArray[T]):
+
+    /** Apply `f` to each element for its side effects.
+      * Note: [U] parameter needed to help scalac's type inference.
+      */
+    def foreach[U](f: T => U): Unit = {
+      val len = xs.length
+      var i = 0
+      while(i < len) {
+        val x = xs(i)
+        if(p(x)) f(x)
+        i += 1
+      }
+    }
+
+    /** Builds a new array by applying a function to all elements of this array.
+      *
+      *  @param f      the function to apply to each element.
+      *  @tparam U     the element type of the returned array.
+      *  @return       a new array resulting from applying the given function
+      *                `f` to each element of this array and collecting the results.
+      */
+    def map[U: ClassTag](f: T => U): IArray[U] = {
+      val b = IArrayBuilder.make[U]
+      var i = 0
+      while (i < xs.length) {
+        val x = xs(i)
+        if(p(x)) b += f(x)
+        i = i + 1
+      }
+      b.result()
+    }
+
+    /** Builds a new array by applying a function to all elements of this array
+      * and using the elements of the resulting collections.
+      *
+      *  @param f      the function to apply to each element.
+      *  @tparam U     the element type of the returned array.
+      *  @return       a new array resulting from applying the given collection-valued function
+      *                `f` to each element of this array and concatenating the results.
+      */
+    def flatMap[U: ClassTag](f: T => IterableOnce[U]): IArray[U] = {
+      val b = IArrayBuilder.make[U]
+      var i = 0
+      while(i < xs.length) {
+        val x = xs(i)
+        if(p(x)) b ++= f(xs(i))
+        i += 1
+      }
+      b.result()
+    }
+
+    def flatMap[BS, U](f: T => BS)(using asIterable: BS => Iterable[U], m: ClassTag[U]): IArray[U] =
+      flatMap[U](x => asIterable(f(x)))
+
+    /** Creates a new non-strict filter which combines this filter with the given predicate. */
+    def withFilter(q: T => Boolean): WithFilter[T] = new WithFilter[T](a => p(a) && q(a), xs)
+
+  end WithFilter
+
+  /** A builder class for immutable arrays.
+  *
+  *  @tparam T    the type of the elements for the builder.
+  */
+  @SerialVersionUID(3L)
+  sealed abstract class IArrayBuilder[T]
```

Comment:
Hum, now that I see it here, I don't think it's the right place for this one, in fact. `ArrayBuilder` is a top-level class in `scala.collection.mutable`, even though `Array` is in `scala.*`. By analogy, `IArrayBuilder` should also be a top-level class in `scala.collection.mutable` (yes, `mutable`, like `StringBuilder`).

Reply:
Moved to `scala.collection.mutable`